Use (and add) more test builders

vdemeester opened the pull request:
Following #310, migrating unit tests in `taskrun_test.go` using those builders. This is a wip as I'm not that good at naming things :man_facepalming: (function names, or, packages…).

This contains some — somewhat heavy — refactoring on those tests only. There is still room for improvement/duplication, especially on some sane default values for some structs and make it even more readable :upside_down_face: .

- should they still be in `./test` package ? 
- should we alias that package (to have sthg like `b.TaskRun(…)`) ?
- or should it bind "dot" import (aka using func as is… ) ? (<- I think it's a bad idea :sweat_smile:)

Friday is refactoring day :angel: :stuck_out_tongue_closed_eyes: 

/cc @bobcatfish @pivotal-nader-ziada

bobcatfish commented on test/builder.go:
```diff
@@ -21,6 +27,36 @@ func Task(name, namespace string, ops ...func(*v1alpha1.Task)) *v1alpha1.Task {
 	return t
 }
 
+func TaskNamespace(namespace string) func(*v1alpha1.Task) {
```
One thought I have is that it would be great to have some docs for the Builders - at the moment the best place for them would be https://github.com/knative/build-pipeline/blob/master/test/README.md#adding-integration-tests but I've been thinking for a while now that we should move that stuff into go docs instead, what do you think?

I think we should definitely include docstrings for each publicly exported function here anyway - and it would be super cool if we actually had some examples too!

We've even got a semi-decent start at https://godoc.org/github.com/knative/build-pipeline/test

vdemeester replied:
Definitely ! I didn't "yet" put godocs but we should definitely have some more for those builder

bobcatfish commented on test/builder.go:
```diff
@@ -21,6 +27,36 @@ func Task(name, namespace string, ops ...func(*v1alpha1.Task)) *v1alpha1.Task {
+	}
+}
+
+func ClusterTask(name string, ops ...func(*v1alpha1.ClusterTask)) *v1alpha1.ClusterTask {
```
definitely could use some comments explaining how params like `ops` work

vdemeester commented:
@bobcatfish updated with way more documentation and examples.
The godoc page look like [this](https://i.imgur.com/qyY2ciG.png) (full-page screenshot, so *way* too big to display inline).

With the doc, the `diff` took quite a beating :sweat_smile:

vdemeester commented:
> One thought I have is that it would be great to have some docs for the Builders - at the moment the best place for them would be https://github.com/knative/build-pipeline/blob/master/test/README.md#adding-integration-tests but I've been thinking for a while now that we should move that stuff into go docs instead, what do you think?
> 
> I think we should definitely include docstrings for each publicly exported function here anyway - and it would be super cool if we actually had some examples too!
> 
> We've even got a semi-decent start at https://godoc.org/github.com/knative/build-pipeline/test

I'll open follow-ups PR to enhance the `knative/build-pipeline/test` godoc (and also the `test/README.md`), but definitely not in this one, the diff is already huge :sweat_smile:

vdemeester commented:
> In file builder_test,go , there are no tests. It has examples on how to use them which is great. The examples in builder_test would be a great README for builder package if you want. If you do create this README then I would suggest adding link for this in doc.go

Good point, updated :wink: 

> I am not fan of adding defaultNamespace into builder. If builders are only limited to unit tests then I think its harmless but in this PR its using in e2e test. For example If there is no namespace set then thats the error surfaced instead of creating object in "foo" namespace which does not exist.

Good point, I wanted to be a default one in `unit tests` to remove some noise, but it makes it really easy to mess up in the e2e tests… I'll change that :wink: 

Updated with : 
- a README
- keeping example file
- adding test coverage ~80%

pivotal-nader-ziada commented:
@vdemeester even though you have awesome commit message, do you think you can squash them a bit in this PR before merging, its just gets hard to keep track in the repo history with having so many 🙃

shashwathi commented on test/builder/pipeline.go:
```diff
@@ -0,0 +1,185 @@
+
+// PipelineRun creates a PipelineRun with default values.
+// Any number of PipelineRun modifier can be passed to transform it.
+func PipelineRun(name string, ops ...PipelineRunOp) *v1alpha1.PipelineRun {
```
How about updating the function definition to 

```
func PipelineRun(name, namespace string, ops ...PipelineRunOp) *v1alpha1.PipelineRun {
pr := &v1alpha1.PipelineRun{
		ObjectMeta: metav1.ObjectMeta{
			Namespace: namespace,
			Name:      name,
		},
....
```

This expects the caller to set name and namespace so there is no assumption about namespace of object. WDYT @vdemeester ?

shashwathi commented on test/builder/pipeline.go:
```diff
@@ -0,0 +1,185 @@
+}
+
+// PipelineRunNamespace sets the namespace of the PipelineRun.
+func PipelineRunNamespace(namespace string) PipelineRunOp {
```
We wouldn't need this function if `PipelineRun` function is updated to above suggestion.

vdemeester commented:
@pivotal-nader-ziada ahah yes I can do that :angel: I kept them separate to ease the PR review but indeed, there is a bit too much, I'll squash them :wink:

shashwathi commented on test/builder/pipeline.go:
```diff
@@ -0,0 +1,185 @@
+
+// PipelineResource creates a PipelineResource with default values.
+// Any number of PipelineResource modifier can be passed to transform it.
+func PipelineResource(name string, ops ...PipelineResourceOp) *v1alpha1.PipelineResource {
```
How about updating `PipelineResource ` definition to

```
func PipelineResource(name, namespace string, ops ... PipelineResourceOp) *v1alpha1. PipelineResource {
```

Like above comment, lets not assume namespace for objects

shashwathi commented on test/builder/pipeline_test.go:
```diff
@@ -0,0 +1,76 @@
+			}},
+		},
+	}
+	if d := cmp.Diff(expectedPipelineResource, pipelineResource); d != "" {
```
Great tests! 👍

vdemeester commented:
@pivotal-nader-ziada @shashwathi Updated :angel:

vdemeester commented:
/test pull-knative-build-pipeline-integration-tests

shashwathi commented on test/builder/README.md:
```diff
@@ -0,0 +1,120 @@
+		myStep,
+	))
+	// … and another one.
+	myOtherTask := tb.Task("my-other-task", "namespace", tb.TaskNamespace("foo"),
```
Is `tb.TaskNamespace("foo")` left over by accident?

vdemeester replied:
indeed :man_facepalming:

shashwathi commented on test/builder/README.md:
````diff
@@ -0,0 +1,120 @@
+```go
+	// Definition
+	type TaskOp func(*v1alpha1.Task)
+	func Task(name string, ops ...TaskOp) *v1alpha1.Task {
````
Docs still refer to the old function signature.

vdemeester replied:
Ah I though I forgot one somewhere :man_facepalming:

shashwathi commented on test/builder/README.md:
```diff
@@ -0,0 +1,120 @@
+		tb.PipelineSpec(tb.PipelineTask("foo", "banana")),
+	)
+	// … and PipelineRun
+	pipelineRun := tb.PipelineRun("pear", tb.PipelineRunNamespace("foo"),
```
One more function missed out here.

vdemeester replied:
Oh boy 😅

vdemeester commented:
> My last suggestion is about file test/builder/builder_test.go. Sorry about so many reviews. It still holds example on how to use builder and not tests. Can you move these examples into doc or update the functions to tests?

Indeed, those are there for the `godoc` examples (inline in the doc), so I can't use `testing.T` in their signature :upside_down_face: I'll rename the file to `examples_test.go` to make it clearer :angel:

shashwathi commented:
/lgtm
